### ISSUE TITLE
Scoping behavior of annotations disagrees with spec

### ISSUE THREAD

**Issue by stereotype441** [dart-lang/language#1790]
The following weird program is accepted by both the analyzer and the CFE:
```dart
class A {
  const A(void Function() x);
}
class C<@A(B) T> {
  static void B() {}
}
main() {
  new C<int>();
}
```
This seems to contradict the spec, which says "Type parameters are declared in the type parameter scope of a class or function" (15 Generics) and "The constant expression given in an annotation is type checked and evaluated in the scope surrounding the declaration being annotated" (16 Metadata).  Together these seems to imply that the reference to `B` in `@A(B)` should be an error, because there is no declaration of `B` in scope.

**Comment by stereotype441** [comment]
Digging a little deeper, the analyzer and CFE seem to be using different mechanisms to perform the (IMHO erroneous) resolution of B.  The analyzer seems to resolve it via an implicit reference to `this` (which seems wrong, because `this` isn't available in annotations), and the CFE seems to resolve it via the class scope (which seems wrong, because the annotation is outside the class body).  You can see the difference if you try analyzing/compiling this example:
```dart
class A {
  const A(void Function() x);
}
class B {}
class C<@A(B) T extends B> {
  static void B() {}
}
main() {
  new C<B>();
}
```
This example is accepted by the CFE, but rejected by the analyzer.  The analyzer gives this error message:
```
error • test.dart:5:12 • The argument type 'Type' can't be assigned to the parameter type 'void Function()'. • argument_type_not_assignable
```

**Comment by stereotype441** [comment]
@eernstg Is my interpretation of the spec correct here?  I believe the first example should be an unresolved identifier error (because there is no `B` in scope at the site of `@A(B)`), and the second example should be a type mismatch error (because the `B` in `@A(B)` should be interpreted as referring to `class B`)

**Comment by stereotype441** [comment]
Note: here is a further illustration that the analyzer is resolving annotations on type arguments via implicit `this`:
```dart
class A {
  const A(void Function() x);
}
class C {
  static void B() {}
}
extension E<@A(B) T> on C {}
main() {}
```
The analyzer reports no error for this, because the `B` in `@A(B)` is resolved to the static method `C.B`.  The CFE rejects it, reporting `Error: Getter not found: 'B'.`

**Comment by eernstg** [comment]
Good catch! ;-)

> the first example should be an unresolved identifier error ..
> the second example should be a type mismatch error

Agreed.

**Comment by stereotype441** [comment]
In discussion with @leafpetersen the other day, we decided that it's extremely unlikely that anyone is taking advantage of this scoping bug, so there's no need to tie the fix to a particular language version to assist folks in migrating.  We can just fix it, provided that we make a note in the changelog.

I just ran an experimental fix (https://dart-review.googlesource.com/c/sdk/+/209940/2) through google3 and found no failures, confirming that no one is taking advantage of this bug anywhere in internal sources.  So I'm going to proceed with the full bug fix.
